File: src/ros2_bridge/mappers/foxglove_msgs/frame_transform.rs
//! Typed mapper for `foxglove_msgs/msg/FrameTransform`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn frame_transform_to_bus(msg: ros_env::foxglove_msgs::msg::FrameTransform) -> crate::foxglove_msgs::msg::v1::FrameTransform {
    crate::foxglove_msgs::msg::v1::FrameTransform {
        timestamp: Some(crate::ros2_bridge::mappers::convert::time_to_timestamp(msg.timestamp)),
        parent_frame_id: crate::ros2_bridge::mappers::convert::from_ros_string(msg.parent_frame_id),
        child_frame_id: crate::ros2_bridge::mappers::convert::from_ros_string(msg.child_frame_id),
        translation: Some(crate::ros2_bridge::mappers::foxglove_msgs::vector3::vector3_to_bus(msg.translation)),
        rotation: Some(crate::ros2_bridge::mappers::foxglove_msgs::quaternion::quaternion_to_bus(msg.rotation)),
    }
}

pub(crate) fn frame_transform_to_ros(bus: crate::foxglove_msgs::msg::v1::FrameTransform) -> ros_env::foxglove_msgs::msg::FrameTransform {
    ros_env::foxglove_msgs::msg::FrameTransform {
        timestamp: crate::ros2_bridge::mappers::convert::timestamp_to_time(bus.timestamp.unwrap_or_default()),
        parent_frame_id: crate::ros2_bridge::mappers::convert::to_ros_string(bus.parent_frame_id),
        child_frame_id: crate::ros2_bridge::mappers::convert::to_ros_string(bus.child_frame_id),
        translation: crate::ros2_bridge::mappers::foxglove_msgs::vector3::vector3_to_ros(bus.translation.unwrap_or_default()),
        rotation: crate::ros2_bridge::mappers::foxglove_msgs::quaternion::quaternion_to_ros(bus.rotation.unwrap_or_default()),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct FoxgloveMsgsFrameTransformMapper;

impl TypedTopicMapper for FoxgloveMsgsFrameTransformMapper {
    type Ros = ros_env::foxglove_msgs::msg::FrameTransform;
    type Bus = crate::foxglove_msgs::msg::v1::FrameTransform;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(frame_transform_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(frame_transform_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/foxglove_msgs/frame_transform.py
"""Generated mapper for `foxglove_msgs/msg/FrameTransform`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.foxglove_msgs.vector3 import vector3_to_bus, vector3_to_ros
from robot_bus.ros2_bridge.mappers.foxglove_msgs.quaternion import quaternion_to_bus, quaternion_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.foxglove_msgs.msg.v1 import FrameTransform as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def frame_transform_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.timestamp = _convert.time_to_timestamp(msg.timestamp)
    bus.parent_frame_id = str(msg.parent_frame_id)
    bus.child_frame_id = str(msg.child_frame_id)
    bus.translation.CopyFrom(vector3_to_bus(msg.translation))
    bus.rotation.CopyFrom(quaternion_to_bus(msg.rotation))
    return bus


def frame_transform_to_ros(bus):
    from foxglove_msgs.msg import FrameTransform as RosMsg

    out = RosMsg()
    out.timestamp = _convert.timestamp_to_time(bus.timestamp)
    out.parent_frame_id = str(bus.parent_frame_id)
    out.child_frame_id = str(bus.child_frame_id)
    out.translation = vector3_to_ros(bus.translation)
    out.rotation = quaternion_to_ros(bus.rotation)
    return out


class FoxgloveMsgsFrameTransformMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from foxglove_msgs.msg import FrameTransform as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return frame_transform_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return frame_transform_to_ros(bus)
